funcx_web_service/authentication/auth_state.py

```diff
@@ -0,0 +1,98 @@
+import typing as t
+
+import globus_sdk
+from flask import abort, current_app, g, request
+
+from funcx_web_service.models.user import User
+
+from .globus_auth import introspect_token
+
+
+class AuthenticationState:
+    """
+    This is a dedicated object for handling authentication.
+
+    It takes in a Globus Auth token and resolve it to a user and various data about that
+    user. It is the "auth_state" object for the application within the context of a
+    request, showing "who" is calling the application (e.g. identity_id) and some
+    information about "how" the call is being made (e.g. scopes).
+
+    For the most part, this should not handle authorization checks, to maintain
+    separation of concerns.
+    """
+
+    # Default scope if not provided in config
+    DEFAULT_FUNCX_SCOPE = (
+        "https://auth.globus.org/scopes/facd7ccc-c5f4-42aa-916b-a0e270e2c2a9/all"
+    )
+
+    def __init__(
+        self, token: t.Optional[str], *, assert_default_scope: bool = True
+    ) -> None:
+        self.funcx_all_scope: str = current_app.config.get(
+            "FUNCX_SCOPE", self.DEFAULT_FUNCX_SCOPE
+        )
+        self.token = token
+
+        self.introspect_data: t.Optional[globus_sdk.GlobusHTTPResponse] = None
+        self.identity_id: t.Optional[str] = None
+        self.username: t.Optional[str] = None
+        self._user_object: t.Optional[User] = None
+        self.scopes: t.Set[str] = set()
+
+        if token:
+            self._handle_token()
+
+    def _handle_token(self) -> None:
+        """Given a token, flesh out the AuthenticationState."""
+        self.introspect_data = introspect_token(t.cast(str, self.token))
+        self.username = self.introspect_data["username"]
+        self.identity_id = self.introspect_data["sub"]
+        self.scopes = set(self.introspect_data["scope"].split(" "))
+
+    @property
+    def user_object(self) -> User:
+        if self._user_object is None:
+            self._user_object = User.resolve_user(self.username)
+        return self._user_object
+
+    @property
+    def is_authenticated(self):
+        return self.identity_id is not None
+
+    def assert_is_authenticated(self):
+        """
+        This tests that is_authenticated=True, and raises an Unauthorized error
+        (401) if it is not.
+        """
+        if not self.is_authenticated:
+            abort(401, "method requires token authenticated access")
+
+    # TODO: determine if a 403 response is appropriate for incorrect scopes
+    # this should possibly be changed to a 401
```

Comment:
While i've seen lot's of arguments about 401s vs 403s and services in the wild are pretty inconsistent, the OAuth spec is pretty consistent in it's use of 401s only when the credentials in the authorization header are somehow invalid or missing and 403s when the the credentials are valid but have insufficient privilege.  Specifically, on the point of missing or insufficient scopes it is [pretty clear](https://datatracker.ietf.org/doc/html/rfc6750#section-3.1):

>    insufficient_scope
         The request requires higher privileges than provided by the
         access token.  The resource server SHOULD respond with the HTTP
         403 (Forbidden) status code and MAY include the "scope"
         attribute with the scope necessary to access the protected
         resource.

That said, I'm sure we aren't 100% spec compliant in other aspects of this service.

Reply:
I wasn't sure which was considered correct, but it struck me as something we should review. This has an even better resolution than I was hoping -- I didn't realize that the spec included this detail.

I'll update the comment to note that RFC 6750 specifically treats insufficient scopes as 403s, instead of the "check this" TODO.